File: PostBPF/BPFGrep.cpp
#include "BPFGrep.h"
#include <iostream>
#define BUFFEREND "**HELALTER***"


BPFGrep::BPFGrep(const std::string& filename) {
    file.open(filename);
    if (!file.is_open()) {
        std::cerr << "Unable to open file: " << filename << std::endl;
    }
}

BPFGrep::~BPFGrep() {
    if (file.is_open()) {
        file.close();
    }
}

std::string BPFGrep::readArgLabel() {
    char ch = '\0';
    std::string arg_type = "";

    if (file.is_open()) {
        while (ch != ':' && file.get(ch)) {
            arg_type += ch;
        }
        std::cout << arg_type << std::endl;
    }
    arg_type.resize(arg_type.size() - 1);
    return arg_type;

}

std::string BPFGrep::readArgPtr() {
    char ch = '\0';
    std::string arg_ptr = "";

    if (file.is_open()) {
        while (ch != '\n' && file.get(ch)) {
            arg_ptr += ch;
        }
        std::cout << arg_ptr << std::endl;
    }
    return arg_ptr;
}

std::string BPFGrep::readBufferContent() {
    char ch = '\0';
    std::string buffer_content = "";
    std::string last13Chars;
    const size_t maxChars = 13;

    if (file.is_open()) {
        while (file.get(ch)) {
            buffer_content += ch;
            last13Chars += ch;

            if (last13Chars.size() > maxChars) {
                last13Chars.erase(0, 1);
            }
            if (isBufferEnd(last13Chars)) {
                break;
            }
        }
        cleanTail(buffer_content);
        std::cout << buffer_content << std::endl;
    }
    return buffer_content;

}

bool BPFGrep::isBufferEnd(const std::string& input) const {
    return input == BUFFEREND;
}

void BPFGrep::cleanTail(std::string& input, const int numCharsToRemove) {
    if (input.size() >= numCharsToRemove) {
        input.resize(input.size() - numCharsToRemove);
    } else {
        std::cerr << "Error: The string is shorter than 13 characters." << std::endl;
    }
}

File: Final/BPFGrep.py
import sys
class BPFGrep:
    def __init__(self, filename, verbose=False):
        self.filename = filename
        self.verbose = verbose  # Class attribute to control printing
        try:
            self.file = open(self.filename, 'r')
        except Exception as e:
            print(f"Unable to open file: {filename}")
            print(e)
            sys.exit(1)

    def __del__(self):
        if self.file and not self.file.closed:
            self.file.close()

    def read_one_arg(self):
        label = self.read_arg_label()
        ptr_addr = self.read_arg_ptr()
        buffer_content = self.read_buffer_content()
        return label, buffer_content, ptr_addr

    # ---------- Helper Functions Below ----------

    def read_arg_label(self):
        arg_type = ""
        while True:
            ch = self.file.read(1)
            if ch == ':' or not ch:
                break
            arg_type += ch
        self._print(arg_type)
        return arg_type

    def read_arg_ptr(self):
        arg_ptr = ""
        while True:
            ch = self.file.read(1)
            if ch == '\n' or not ch:
                break
            arg_ptr += ch
        self._print(arg_ptr)
        return self.to_int(arg_ptr)

    def to_int(self, input_str):
        try:
            return int(input_str)
        except ValueError as e:
            self._print(f"Invalid argument: {input_str}")
            return 0

    def read_buffer_content(self):
        buffer_content = ""
        last13_chars = ""
        max_chars = 13

        while True:
            ch = self.file.read(1)
            if not ch:
                break
            buffer_content += ch
            last13_chars += ch

            if len(last13_chars) > max_chars:
                last13_chars = last13_chars[1:]

            if self.is_buffer_end(last13_chars):
                break
        buffer_content = self.clean_tail(buffer_content, max_chars)
        self._print(buffer_content)
        return buffer_content

    def is_buffer_end(self, input_str):
        return input_str == "**HELALTER***"

    def clean_tail(self, input_str, num_chars_to_remove):
        if len(input_str) >= num_chars_to_remove:
            return input_str[:-num_chars_to_remove]
        else:
            self._print("Error: The string is shorter than 13 characters.")
            return input_str

    # Private helper method to control printing
    def _print(self, message):
        if self.verbose:
            print(message)